### PR TITLE
Reorganize credit termstructures

### PR DESCRIPTION
**Pull request by thrasibule**
- PiecewiseDefault curve now depends on DefaultTermStructure
- Reorganize and dd some missing methods
- add another type of termstructure (FlatHazardRate) which is quite useful

### CODE REVIEW

File: quantlib/termstructures/credit/flat_hazard_rate.pyx [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    def __init__(self, int settlement_days,  Calendar calendar not None,
+                 double hazard_rate, DayCounter day_counter not None):
+        self._thisptr = shared_ptr[_dts.DefaultProbabilityTermStructure](
+            new _fhr.FlatHazardRate(settlement_days, deref(calendar._thisptr),
+                                    hazard_rate, deref(day_counter._thisptr)))
+
+    @classmethod
+    def from_reference_date(cls, Date reference_date not None, double hazard_rate,
+                            DayCounter day_counter not None):
+        cdef FlatHazardRate instance = cls.__new__(cls)
```
Comment (dpinte) [inline quantlib/termstructures/credit/flat_hazard_rate.pyx]: why not just `cls()` in place of `cls.__new__(cls)` ?

**Comment by dpinte** [comment]
Thanks @thrasibule . One minor comment and then ready to merge.

File: quantlib/termstructures/credit/flat_hazard_rate.pyx (hunk 1 above)
Comment (thrasibule) [inline quantlib/termstructures/credit/flat_hazard_rate.pyx]: Calling cls() would not work cause _init__ takes some required arguments. It's a bit ugly, but it's a way to create an instance of type FlatHazardRate without calling __init__ on it. (see http://docs.cython.org/src/userguide/extension_types.html?highlight=__new__#fast-instantiation).

Comment (dpinte) [inline quantlib/termstructures/credit/flat_hazard_rate.pyx]: Got you! Ok, that's fine then.

**Pull request merged**